azerothcore/azerothcore-wotlk
# chore: Cmake vars for bin/conf installation pathes

This PR is useful for people who want to structure their directories logically which is probably required if you have more than one realm (example: 1 production realm, 1 development realm, or multilingual realms etc...). It also improves the default structure in my opinion.

## CHANGES PROPOSED:

- Display where the files are going to be installed in the cmake output (atleast on linux in command line)
- Add extra variables for much more flexibility over where your files are installed (see below)

Here are the variables you can set now:

```
INSTALL_PATH: Set wherever you want to install the server files after compilation - REQUIRED
AUTHSERVER_BIN_DIR: Set the directory for the authserver's binary (relative to INSTALL_PATH)
WORLDSERVER_BIN_DIR: Set the directory for the worldserver's binary (relative to INSTALL_PATH)
MODULES_CONF_DIR: Set the directory for the modules' configuration files (relative to INSTALL_PATH)
```

Basically `*_DIR` refers to the relative path where `INSTALL_PATH` is the root directory, while `*_PATH` refers to the absolute path. `*_DIR` variables exist for convenience.

Shorthand variables:

```
SERVER_BIN_DIR: Shorthand for AUTHSERVER_BIN_DIR + WORLDSERVER_BIN_DIR
SERVER_BIN_PATH: Shorthand for AUTHSERVER_BIN_PATH + WORLDSERVER_BIN_PATH
SERVER_CONF_PATH: Shorthand for AUTHSERVER_CONF_PATH + WORLDSERVER_CONF_PATH
```

Here are more variables you can set, but you will probably not feel the need to use them at all:

```
PREFIX: This is the same variable as INSTALL_PATH but it's just kept for people who are used to the old variable name, it is deprecated and should be removed in the future.
AUTHSERVER_BIN_PATH: Change the full path to the authserver binary (absolute path)
WORLDSERVER_BIN_PATH: Change the full path to the worldserver binary (absolute path)
AUTHSERVER_CONF_PATH: Change the full path to the authserver config files (absolute path)
WORLDSERVER_CONF_PATH: Change the full path to the worldserver config files (absolute path)
MODULES_CONF_PATH: Change the full path to the modules configuration files (absolute path)
```

Note: there is no `AUTHSERVER_CONF_BIN` nor `WORLDSERVER_CONF_BIN` variables.



### Examples :

**Default installation**

Cmake variables:

```cmake
INSTALL_PATH = /path/to/my/azerothcore/
```

Files installed in:

```
authserver binary -> /path/to/my/azerothcore/authserver.exe
authserver conf file -> /path/to/my/azerothcore/authserver.conf.dist
worldserver -> /path/to/my/azerothcore/worldserver.exe
worldserver conf file -> /path/to/my/azerothcore/worldserver.conf.dist
modules conf files -> /path/to/my/azerothcore/modules_conf/...
```

**Example 1**

Cmake variables:

```cmake
INSTALL_PATH = /path/to/my/azerothcore/
AUTHSERVER_BIN_DIR = authserver
MODULES_CONF_PATH = /path/superconf/
```

Files installed in:

```
authserver binary -> /path/to/my/azerothcore/authserver/authserver.exe
authserver conf file -> /path/to/my/azerothcore/authserver/authserver.conf.dist
modules conf files -> /path/superconf/...
```

**Example 2**

Cmake variables:

```cmake
INSTALL_PATH = /path/to/my/azerothcore/
AUTHSERVER_BIN_PATH = /another_path/authserver/
AUTHSERVER_CONF_PATH = /another_path/authserver/
MODULES_CONF_DIR = test_conf/subdir/
```

Files installed in:

```
authserver binary -> /another_path/authserver/authserver.exe
authserver conf file -> /another_path/authserver/authserver.conf.dist
modules conf files -> /path/to/my/azerothcore/test_conf/subdir/...
```


## ISSUES ADDRESSED:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/1519

## TESTS PERFORMED:
- It compiles fine on linux
- I tried a combination of the variables and it worked

## HOW TO TEST THE CHANGES:

- Clean your "build/" directory
- Run cmake with one/several of the new CMAKE options
- Check the cmake output, in the middle (on linux in command line) it will list where your files are going to be installed starting with:
```
-- Detected 64-bit platform                                                                                                                         
-- INSTALL: Installing files in /path/...
```
- Compile and install
- Verify it installed files in the right location

##### KNOWN ISSUES AND TODO LIST:

- [ ] Test on windows (i don't have windows to test)
- [ ] Add the info to the wiki page

## Target branch(es):


 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR
